src/index/StringSortComparator.h

```diff
@@ -0,0 +1,462 @@
+        ignorePunctuationAtFirstLevel ? UCOL_SHIFTED : UCOL_NON_IGNORABLE;
+
+    if (_icuLocale.isBogus() == TRUE) {
+      throw std::runtime_error("Could not create locale with language " + lang +
+                               " and Country " + country);
+    }
+    setupCollators();
+    setIgnorePunctuationOnFirstLevels(_ignorePunctuationStatus);
+  }
+
+  // ____________________________________________________________________
+  LocaleManager& operator=(const LocaleManager& other) {
+    if (this == &other) return *this;
+    _icuLocale = other._icuLocale;
+    _ignorePunctuationStatus = other._ignorePunctuationStatus;
+    setupCollators();
+    setIgnorePunctuationOnFirstLevels(_ignorePunctuationStatus);
+    return *this;
+  }
+
+  // _____________________________________________________________________
+  [[nodiscard]] int compare(std::string_view a, std::string_view b,
+                            const Level level) const {
+    UErrorCode err = U_ZERO_ERROR;
+    auto idx = static_cast<uint8_t>(level);
+    auto res = compToInd(
+        _collator[idx]->compareUTF8(toStringPiece(a), toStringPiece(b), err));
+    raise(err);
+    return res;
+  }
+
+  // _____________________________________________________________________
+  static int compare(WeightString a, WeightString b,
+                     [[maybe_unused]] const Level) {
+    return std::strcmp(a.get().c_str(), b.get().c_str());
+  }
+
+  // transform a UTF-8 string into a weight string that can be compared using
+  // std::strcmp. needed because ICU only works on utf16 and does not create a
+  // std::string.
+  [[nodiscard]] WeightString getSortKey(std::string_view s,
+                                        const Level level) const {
+    auto utf16 = icu::UnicodeString::fromUTF8(toStringPiece(s));
+    auto& col = *_collator[static_cast<uint8_t>(level)];
+    auto sz = col.getSortKey(utf16, nullptr, 0);
+    WeightString finalRes;
+    std::string& res = finalRes.get();
+    res.resize(sz);
+    static_assert(sizeof(uint8_t) == sizeof(std::string::value_type));
+    sz = col.getSortKey(utf16, reinterpret_cast<uint8_t*>(res.data()),
+                        res.size());
+    AD_CHECK(sz == static_cast<decltype(sz)>(
+                       res.size()));  // this is save by the way we obtained sz
+    // since this is a c-api we still have a trailing '\0'. Trimming this is
+    // necessary for the prefix range to work correct.
+    res.resize(res.size() - 1);
+    return finalRes;
+  }
+
+      [[nodiscard]] std::string getLowercaseUtf8(const std::string& s) const {
+    auto utf16 = icu::UnicodeString::fromUTF8(s);
+    utf16 = utf16.toLower(_icuLocale);
+    std::string res;
+    utf16.toUTF8String(res);
+    return res;
+  }
+
+  /// get a prefix of length prefixLength of the UTF8 string (or shorter, if the
+  /// string contains less codepoints This counts utf-8 codepoints correctly and
+  /// returnsa UTF-8 string referring to the Prefix and the number of utf
```

Comment:
missing space in "returnsa". Also I think the correct term is Unicode codepoints not utf codepoints